retworkx-core/src/traversal/dijkstra_visit.rs

```diff
@@ -0,0 +1,324 @@
+///
+/// Find the shortest path from vertex 0 to 5, and exit the visit as soon as
+/// we reach the goal vertex.
+///
+/// ```
+/// use retworkx_core::petgraph::prelude::*;
+/// use retworkx_core::petgraph::graph::node_index as n;
+/// use retworkx_core::petgraph::visit::Control;
+///
+/// use retworkx_core::traversal::{DijkstraEvent, dijkstra_search};
+///
+/// let gr: Graph<(), ()> = Graph::from_edges(&[
+///     (0, 1), (0, 2), (0, 3), (0, 4),
+///     (1, 3),
+///     (2, 3), (2, 4),
+///     (4, 5),
+/// ]);
+///
+/// // record each predecessor, mapping node → node
+/// let mut predecessor = vec![NodeIndex::end(); gr.node_count()];
+/// let start = n(0);
+/// let goal = n(5);
+/// dijkstra_search(
+///     &gr,
+///     Some(start),
+///     |edge| -> Result<usize, ()> { 
+///         Ok(1) 
+///     },
+///     |event| {
+///         match event {
+///             DijkstraEvent::Discover(v, _) => {
+///                 if v == goal {
+///                     return Control::Break(v);
+///                 }   
+///             },
+///             DijkstraEvent::EdgeRelaxed(u, v, _) => {
+///                 predecessor[v.index()] = u;
+///             },
+///             _ => {}
+///         };
+///
+///         Control::Continue
+///     },
+/// ).unwrap();
+///
+/// let mut next = goal;
+/// let mut path = vec![next];
+/// while next != start {
+///     let pred = predecessor[next.index()];
+///     path.push(pred);
+///     next = pred;
+/// }
+/// path.reverse();
+/// assert_eq!(&path, &[n(0), n(4), n(5)]);
+/// ```
+pub fn dijkstra_search<G, I, F, K, E, H, C>(
+    graph: G,
+    starts: I,
+    mut edge_cost: F,
+    mut visitor: H,
+) -> Result<C, E>
+where
+    G: IntoEdges + Visitable,
+    G::NodeId: Eq + Hash,
+    I: IntoIterator<Item = G::NodeId>,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
+    K: Measure + Copy,
+    H: FnMut(DijkstraEvent<G::NodeId, &G::EdgeWeight, K>) -> C,
+    C: ControlFlow,
+{
+    let visited = &mut graph.visit_map();
+
+    for start in starts {
+        // `dijkstra_visitor` returns a "signal" to either continue or exit early
+        // but it never "prunes", so we use `unreachable`.
+        try_control!(
+            dijkstra_visitor(
+                graph,
+                start,
+                &mut edge_cost,
+                &mut visitor,
+                visited
+            ),
+            unreachable!()
+        );
+    }
+
+    Ok(C::continuing())
+}
+
+pub fn dijkstra_visitor<G, F, K, E, V, C>(
+    graph: G,
+    start: G::NodeId,
+    mut edge_cost: F,
+    mut visitor: V,
+    visited: &mut G::Map,
+) -> Result<C, E>
+where
+    G: IntoEdges + Visitable,
+    G::NodeId: Eq + Hash,
+    F: FnMut(G::EdgeRef) -> Result<K, E>,
+    K: Measure + Copy,
+    V: FnMut(DijkstraEvent<G::NodeId, &G::EdgeWeight, K>) -> C,
+    C: ControlFlow,
+{
+    if visited.is_visited(&start) {
+        return Ok(C::continuing());
+    }
+
+    let mut scores = HashMap::new();
```

Comment:
If we use `HashMap` instead of `DictMap`, we are changing the behavior fo the current implementation. For consitency and determinism, I think `DictMap` is a better choice

Reply:
We do not iterate over `scores` or return it, so `DictMap`/`HashMap` has no difference here.

Reply:
`scores` is returned to Python on our top-level functions, so there is an implicit iteration when converting to a Python dictionary

Reply:
Here we are in `traversal/dijkstra_visit.rs`. We return `scores` in `shortest_path/dijkstra.rs` where it's still a `DictMap`. https://github.com/Qiskit/retworkx/blob/dbbd9c9549920298c9e167cec37fac6f6fd79a12/retworkx-core/src/shortest_path/dijkstra.rs#L111

Reply:
Indeed, we do not return scores here